Add support for `node:*` imports with the `nodejs_compat` compatibility flag

GregBrimble opened the pull request:
What this PR solves / how to test:

This PR does two things:
1. Renames the `--node-compat` CLI/config option _internally_ to `legacyNodeCompat`. Nothing changes externally. You still invoke it with `--node-compat` or `node_compat`. We maybe want to think about moving this to `experimental` as we stabilize the`dev()` API, but that discussion can happen elsewhere.
2. Marks any imports prefixed with `node:` as external when bundling with the `nodejs_compat` compatibility flag.

Associated docs issues/PR:

- https://github.com/cloudflare/workerd/pull/208
- https://github.com/cloudflare/miniflare/issues/473

Author has included the following, where applicable:

- [x] Tests (need a test to check the `publish --dry-run` and `pages functions build` output)
- [x] Changeset

Reviewer has performed the following, where applicable:

- [x] Checked for inclusion of relevant tests
- [x] Checked for inclusion of a relevant changeset
- [x] Checked for creation of associated docs updates
- [ ] Manually pulled down the changes and spot-tested

Fixes # [insert issue number].

GregBrimble commented:
We need to add support for this compatibility flag in Miniflare (https://github.com/cloudflare/miniflare/issues/473) but once [the workerd PR](https://github.com/cloudflare/workerd/pull/208) is merged and released, this'll be available with `--experimental-local`. Difficult (read impossible) to test this properly without either of these in place.

JacobMGEvans reviewed (dismissed):
Could we get tests for preventing users from running the legacy flag and the runtime flag together?

GregBrimble commented:
There are checks for exclusivitity throughout?

- https://github.com/cloudflare/wrangler2/pull/2539/files#diff-91077b2ba8090bca91e063f04c1e862495858f9d4033fa7a8d4e34d9fcdb34e3R346
- https://github.com/cloudflare/wrangler2/pull/2539/files#diff-c90ff1ca6d929baaafb7d67d6d6ccd573593762adf4e7a9b6917343ceeceb68bR309
- https://github.com/cloudflare/wrangler2/pull/2539/files#diff-4aff2d0b1b481b9ead6cbf181ebf7298c13ecccd9644f5e8d4402663f7ff1d5aR134
- https://github.com/cloudflare/wrangler2/pull/2539/files#diff-ee7e6a6ed4db87e45fe346d2922bd6b7eb68485222e23852bed880f5f6468e4fR32
- https://github.com/cloudflare/wrangler2/pull/2539/files#diff-09e3e18998ee74fff220a99061a72ddc6d412dd8fd0110d75a8d20780657d15eR796

JacobMGEvans commented:
> There are checks for exclusivitity throughout?
> 
> * https://github.com/cloudflare/wrangler2/pull/2539/files#diff-91077b2ba8090bca91e063f04c1e862495858f9d4033fa7a8d4e34d9fcdb34e3R346
> * https://github.com/cloudflare/wrangler2/pull/2539/files#diff-c90ff1ca6d929baaafb7d67d6d6ccd573593762adf4e7a9b6917343ceeceb68bR309
> * https://github.com/cloudflare/wrangler2/pull/2539/files#diff-4aff2d0b1b481b9ead6cbf181ebf7298c13ecccd9644f5e8d4402663f7ff1d5aR134
> * https://github.com/cloudflare/wrangler2/pull/2539/files#diff-ee7e6a6ed4db87e45fe346d2922bd6b7eb68485222e23852bed880f5f6468e4fR32
> * https://github.com/cloudflare/wrangler2/pull/2539/files#diff-09e3e18998ee74fff220a99061a72ddc6d412dd8fd0110d75a8d20780657d15eR796

Correct, I would like to see tests that as a proof/coverage of that check functionality.

danbars commented:
What is the status of this PR?
Will it enable the use of `async_hooks` in worker scripts?

Currently I get this error when trying to use it:
`The package "async_hooks" wasn't found on the file system but is built into node. Are you trying to bundle for node? You can use "platform: 'node'" to do that, which will remove this error.`

Is this PR related / will enable this workerd PR: https://github.com/cloudflare/workerd/pull/208 ?

GregBrimble commented:
@danbars , still a work-in-progress! Please bear with us.

irvinebroque commented on .changeset/late-bulldogs-help.md:
```diff
@@ -0,0 +1,7 @@
+
+feat: Add support for the `nodejs_compat` Compatibility Flag when bundling a Worker with Wrangler
+
+This new Compatibility Flag is incompatible with the legacy `--node-compat` CLI argument and `node_compat` configuration option. If you want to use the new runtime Node.js compatibility features, please remove the `--node-compat` argument from your CLI command or your config file.
```
I'm sure we've thought about this a bit already, so don't want to make others rehash past discussion, but I'm curious what we're trying to guard against?

In my mind, if both flags are enabled, then that should mean:

1. Node imports that are built into workerd aren't polyfilled (new `nodejs_compat`)
2. Node imports that aren't built into workerd are polyfilled (old `node_compat`)

...such that people who today need a mix of node APIs that are / aren't in workerd can start using our built-in support right now, and don't have to wait until all the node APIs they need are available in workerd, to make the switch.

GregBrimble replied:
The polyfills are just generally sub-par. We should deprecate ASAP imo, and this is a totally fair forcing function. Just taking advantage of it to encourage people to move over to the proper `nodejs_compat` compatibility flag.

danbars commented:
Hi @GregBrimble, sorry for nagging, I simply have a module that is pending on this PR - using `async_hooks` from node. 
I'll be happy to know what is the expected timeline for this.
If this is a few weeks I guess I'll wait. Otherwise I'll probably develop an alternative solution that does not have this dependency.

GregBrimble commented:
@JacobMGEvans: rebased and added a couple of tests for that now:

- https://github.com/cloudflare/workers-sdk/pull/2539/files#diff-8f31996226fa2e6948e321edadd5c6e1c8eca2ea47ec7662d0b9b41dd8c5dd17R1579
- https://github.com/cloudflare/workers-sdk/pull/2539/files#diff-14ab3d4c8d99fc0fd46940e551843a07935a06e9153869e48b1c4554efcecb97R6782

JacobMGEvans reviewed (approved):
Looks great. I like the "conjunction" messaging and I appreciate the tests around trying to use them together.

irvinebroque commented:
Should this work with Pages? See https://discord.com/channels/595317990191398933/1024343493251973190/1084637493787373630
